packages/prisma-client/prisma/migrations/20221201075120_user-props/migration.ts

```diff
@@ -0,0 +1,139 @@
+import { z } from "zod";
+import { PrismaClient } from "./client";
+
+const baseUserProps = {
+  id: z.string(),
+  prop: z.string(),
+  required: z.optional(z.boolean()),
+};
+
+export const UserDbProp = z.discriminatedUnion("type", [
+  z.object({
+    ...baseUserProps,
+    type: z.literal("number"),
+    value: z.number(),
+  }),
+  z.object({
+    ...baseUserProps,
+    type: z.literal("string"),
+    value: z.string(),
+  }),
+  z.object({
+    ...baseUserProps,
+    type: z.literal("boolean"),
+    value: z.boolean(),
+  }),
+  z.object({
+    ...baseUserProps,
+    type: z.literal("asset"),
+    // In database we hold asset_id
+    value: z.string(),
+  }),
+]);
+
+const UserDbProps = z.array(UserDbProp);
+type UserDbProp = z.infer<typeof UserDbProp>;
+
+const OutdatedProp = z.object({
+  ...baseUserProps,
+  value: z.optional(z.union([z.string(), z.number(), z.boolean()])),
+  asset_id: z.optional(z.string()),
+  assetId: z.optional(z.string()),
+  type: z.optional(z.string()),
+});
+
+const OutdatedProps = z.array(OutdatedProp);
+
+export default () => {
+  const client = new PrismaClient({
+    // Uncomment to see the queries in console as the migration runs
+    // log: ["query", "info", "warn", "error"],
+  });
+  return client.$transaction(
+    async (prisma) => {
+      const allInstanceProps = await prisma.instanceProps.findMany({});
+
+      for (const instanceProps of allInstanceProps) {
+        const rawProps = JSON.parse(instanceProps.props);
+        const props = OutdatedProps.parse(rawProps);
```

Comment:
This may throw when we run migration in production as it's possible that we have invalid data there.

I thinks we should do one of these:
1. Import prod db locally and try applying the migration
2. Wrap this in try..catch and delete row or something so migration would complete anyway
3. Parse all rows in a separate loop before the main loop where you write to the DB. Catch all parse errors and collect them in an array. Before the second loop, if array is not empty, throw an error with a message that would help us decide how to fix prod DB (Migration will fail, but with a useful error)

Reply:
Start testing locally, 
- [ ] - copy props
```
CREATE EXTENSION dblink;

INSERT INTO "public"."InstanceProps"("id","instanceId","treeId","props")
SELECT * FROM dblink (
'dbname=postgres port=5432 host=prodhost user=postgres password=prodpasswd',
'select id, "instanceId", "treeId", "props" from "InstanceProps"'
) AS t(
    id text,
    "instanceId" text,
    "treeId" text,
    props text
)
ON CONFLICT DO NOTHING;
```

- [ ] Test

Reply:
Thank you, really had timeout on production data